### PR TITLE
Convert account controller to ts

### PR DESCRIPTION
**Pull request by Miodec**
#2727 needs to be merged first

Part of #2431

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -99,6 +99,8 @@ backend/.env
 static/adtest.html
 backend/migrationStats.txt
 
+frontend/src/scripts/constants/firebase-config.ts
```
Comment (Bruception) [inline .gitignore:102]: Adding these again? lol

File: frontend/src/scripts/controllers/account-controller.ts [hunk 2]
```diff
@@ -40,17 +42,22 @@ import {
...
 export const gmailProvider = new GoogleAuthProvider();
 
-export function sendVerificationEmail() {
+export function sendVerificationEmail(): void {
   Loader.show();
-  let cu = Auth.currentUser;
-  sendEmailVerification(cu)
+  const user = Auth.currentUser;
+  if (user === null) return;
+  sendEmailVerification(user)
```
Comment (Bruception) [inline frontend/src/scripts/controllers/account-controller.ts:57]: Can we use async/await here instead?

Comment (Bruception) [reply frontend/src/scripts/controllers/account-controller.ts:57]: nvm

File: frontend/src/scripts/controllers/account-controller.ts [hunk 3]
```diff
@@ -70,9 +77,10 @@ export async function getDataAndInit() {
     LoadingPage.updateText("Downloading user data...");
     await LoadingPage.showBar();
     await DB.initSnapshot();
-  } catch (e) {
+  } catch (error) {
+    const e = error as { message: string; responseCode: number };
     AccountButton.loading(false);
-    if (e?.response?.status === 429) {
+    if (e?.responseCode === 429) {
```
Comment (Bruception) [inline frontend/src/scripts/controllers/account-controller.ts:83]: Why `e?`?

File: frontend/src/scripts/controllers/account-controller.ts [hunk 4]
```diff
@@ -103,8 +111,8 @@ export async function getDataAndInit() {
   ResultFilters.loadTags(snapshot.tags);
 
   Promise.all([Misc.getLanguageList(), Misc.getFunboxList()]).then((values) => {
-    let languages = values[0];
-    let funboxModes = values[1];
+    const languages = values[0];
```
Comment (Bruception) [inline frontend/src/scripts/controllers/account-controller.ts:114]: I prefer array destructured: `const [languages, funboxModes] = values;`

File: frontend/src/scripts/controllers/account-controller.ts [hunk 5]
```diff
@@ -115,48 +123,39 @@ export async function getDataAndInit() {
     ResultFilters.load();
   });
 
-  let user = Auth.currentUser;
   if (!snapshot.name) {
     //verify username
-    if (Misc.isUsernameValid(user.name)) {
```
Comment (Bruception) [inline frontend/src/scripts/controllers/account-controller.ts:121]: Is it okay to remove this logic?

File: frontend/src/scripts/controllers/account-controller.ts [hunk 6]
```diff
@@ -492,7 +503,11 @@ export async function removeGoogleAuth() {
       await reauthenticateWithPopup(user, gmailProvider);
     } catch (e) {
       Loader.hide();
-      return Notifications.add(e.message, -1);
+      let message = "Failed to reauthenticate";
```
Comment (Bruception) [inline frontend/src/scripts/controllers/account-controller.ts:506]:
```js
const note = e instanceof Error ? `: ${e.message}` : "";
const message = `Failed to reauthenticate${note}`;
```

File: frontend/src/scripts/controllers/account-controller.ts [hunk 7]
```diff
@@ -515,21 +530,29 @@ export async function removeGoogleAuth() {
...
+  if (user === null) return;
   if (
     user.providerData.find((provider) => provider.providerId === "google.com")
   ) {
     try {
       await reauthenticateWithPopup(user, gmailProvider);
     } catch (e) {
       Loader.hide();
-      return Notifications.add("Could not reauthenticate: " + e.message, -1);
+      let message = "Failed to reauthenticate";
```
Comment (Bruception) [inline frontend/src/scripts/controllers/account-controller.ts:547]: Duplicate logic, maybe consider a function.

File: frontend/src/scripts/controllers/account-controller.ts [hunk 8]
```diff
@@ -606,7 +633,7 @@ async function signUp() {
     const signInResponse = await Ape.users.create(
       nname,
       email,
-      createdAuthUser.user.id
+      createdAuthUser.user.uid
```
Comment (Bruception) [inline frontend/src/scripts/controllers/account-controller.ts:636]: :O Bug or API change?

File: frontend/src/scripts/controllers/account-controller.ts [hunk 9]
```diff
@@ -644,25 +671,23 @@ async function signUp() {
...
     }
-    let txt;
-    if (e.response) {
-      txt =
-        e.response.data.message ||
-        e.response.status + " " + e.response.statusText;
-    } else {
-      txt = e.message;
+    console.log(e);
+    let message = "Failed to create account";
```
Comment (Bruception) [inline frontend/src/scripts/controllers/account-controller.ts:675]: Def consider a function: `getErrorWithNote`

File: frontend/src/scripts/controllers/account-controller.ts (hunk 8 above)
Comment (Miodec) [reply frontend/src/scripts/controllers/account-controller.ts:636]: Bug, there is no id field on the user

File: .gitignore (hunk 1 above)
Comment (Miodec) [reply .gitignore:102]: What do you mean exactly - should be there

File: frontend/src/scripts/controllers/account-controller.ts (hunk 3 above)
Comment (Miodec) [reply frontend/src/scripts/controllers/account-controller.ts:83]: Removed

File: frontend/src/scripts/controllers/account-controller.ts (hunk 4 above)
Comment (Miodec) [reply frontend/src/scripts/controllers/account-controller.ts:114]: Done

File: frontend/src/scripts/controllers/account-controller.ts (hunk 5 above)
Comment (Miodec) [reply frontend/src/scripts/controllers/account-controller.ts:121]: Honestly im not sure

File: .gitignore (hunk 1 above)
Comment (Bruception) [reply .gitignore:102]: > What do you mean exactly - should be there

Look at the file in your branch

Comment (Miodec) [reply .gitignore:102]: Sorry still not sure what youre saying (smooth brain moment)

Comment (Miodec) [reply .gitignore:102]: I see it now

File: frontend/src/scripts/controllers/account-controller.ts (hunk 5 above)
Comment (Miodec) [reply frontend/src/scripts/controllers/account-controller.ts:121]: Im pretty sure this is needed for when the Google Sign in process doesn't complete and the user document doesn't have a name stored.

File: frontend/src/scripts/controllers/account-controller.ts [hunk 10]
```diff
@@ -1,3 +1,5 @@
+//@ts-nochec
```
Comment (Bruception) [inline frontend/src/scripts/controllers/account-controller.ts:1]: Whats up with this comment?

File: frontend/src/scripts/controllers/account-controller.ts (hunk 5 above)
Comment (Bruception) [reply frontend/src/scripts/controllers/account-controller.ts:121]: > Im pretty sure this is needed for when the Google Sign in process doesn't complete and the user document doesn't have a name stored.

If it is needed why has it been removed?

File: frontend/src/scripts/controllers/account-controller.ts (hunk 10 above)
Comment (Miodec) [reply frontend/src/scripts/controllers/account-controller.ts:1]: Whoops. I was for testing when converting to ts. Removed.

**Review by Bruception — APPROVED** [review]
LGTM IIW

**Pull request merged**